libbpf-tools/offcputime.c

```diff
@@ -0,0 +1,350 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2021 Wenbo Zhang
+//
+// Based on offcputime(8) from BCC by Brendan Gregg.
+// 19-Mar-2021   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "offcputime.h"
+#include "offcputime.skel.h"
+#include "trace_helpers.h"
+
+static struct env {
+	pid_t pid;
+	pid_t tid;
+	bool user_threads_only;
+	bool kernel_threads_only;
+	int stack_storage_size;
+	int perf_max_stack_depth;
+	__u64 min_block_time;
+	__u64 max_block_time;
+	long state;
+	int duration;
+	bool verbose;
+} env = {
+	.pid = -1,
+	.tid = -1,
+	.stack_storage_size = 1024,
+	.perf_max_stack_depth = 127,
+	.min_block_time = 1,
+	.max_block_time = -1,
+	.state = -1,
+	.duration = 99999999,
+};
+
+static volatile bool exiting;
+
+const char *argp_program_version = "offcputime 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Summarize off-CPU time by stack trace.\n"
+"\n"
+"USAGE: offcputime [--help] [-p PID | -u | -k] [-m MIN-BLOCK-TIME] "
+"[-M MAX-BLOCK-TIME] [--state] [--perf-max-stack-depth] [--stack-storage-size] "
+"[duration]\n"
+"EXAMPLES:\n"
+"    offcputime             # trace off-CPU stack time until Ctrl-C\n"
+"    offcputime 5           # trace for 5 seconds only\n"
+"    offcputime -m 1000     # trace only events that last more than 1000 usec\n"
+"    offcputime -M 10000    # trace only events that last less than 10000 usec\n"
+"    offcputime -p 185      # only trace threads for PID 185\n"
+"    offcputime -t 188      # only trace thread 188\n"
+"    offcputime -u          # only trace user threads (no kernel)\n"
+"    offcputime -k          # only trace kernel threads (no user)\n";
+
+#define OPT_PERF_MAX_STACK_DEPTH	1 /* --pef-max-stack-depth */
+#define OPT_STACK_STORAGE_SIZE		2 /* --stack-storage-size */
+#define OPT_STATE			3 /* --state */
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Trace this PID only" },
+	{ "tid", 't', "TID", 0, "Trace this TID only" },
+	{ "user-threads-only", 'u', NULL, 0,
+	  "User threads only (no kernel threads)" },
+	{ "kernel-threads-only", 'k', NULL, 0,
+	  "Kernel threads only (no user threads)" },
+	{ "perf-max-stack-depth", OPT_PERF_MAX_STACK_DEPTH,
+	  "PERF-MAX-STACK-DEPTH", 0, "the limit for both kernel and user "
+	  "frames (deault 127)" },
```

Comment:
please don't split strings

also typo: default

and "stack traces"?

Reply:
Fixed, thanks. I will submit a pr later to fix all string split problems in other tools. I understand the reason for this from your last pr comment. This time is too busy to update this pr.